contracts/Staking.sol

```diff
@@ -458,66 +446,110 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
+
+        // if lock gets down to zero, unlock
+        if (lock_.amount == _amount) {
+            // it comes from transferFromLock, which uses msg.sender as _lockManager
+            _unsafeUnlock(_accountAddress, _lockManager);
         } else {
-            lock_.amount = lock_.amount.sub(_amount);
-            // if lock gets down to zero, unlock
-            if (lock_.amount == 0) {
-                _unlock(_accountAddress, _lockId);
-            }
+            // no need for SafeMath: checked in require above
+            lock_.amount = lock_.amount - _amount;
+
+            // update total
+            // no need for SafeMath: totalLocked must be greater or equal than lock._amount
+            account.totalLocked = account.totalLocked - _amount;
         }
     }
 
-    function _transfer(address _from, uint256 _fromLockId, address _to, uint256 _toLockId, uint256 _amount) internal {
+    function _transfer(address _from, address _fromLockManager, address _to, address _toLockManager, uint256 _amount) internal {
         // transferring 0 staked tokens is invalid
         require(_amount > 0, ERROR_AMOUNT_ZERO);
 
-        // first valid lock starts at 1, so _toLockId = 0 means no lock
-        if (_toLockId > 0) {
-            _updateActiveLockAmount(_to, _toLockId, _amount, true);
+        // _toLockManager = 0 means no lock
+        if (_toLockManager != address(0)) {
+            _increaseActiveLockAmount(_to, _toLockManager, _amount);
         }
 
         // update stakes
         _modifyStakeBalance(_from, _amount, false);
         _modifyStakeBalance(_to, _amount, true);
 
-        emit StakeTransferred(_from, _fromLockId, _amount, _to, _toLockId);
+        emit StakeTransferred(_from, _fromLockManager, _amount, _to, _toLockManager);
+    }
+
+    /**
+     * @notice Get the staked but unlocked amount of tokens by `_accountAddress`
+     * @param _accountAddress Owner of the staked but unlocked balance
+     * @return Amount of tokens staked but not locked by given account
+     */
+    function _unlockedBalanceOf(address _accountAddress) internal view returns (uint256) {
+        uint256 unlockedTokens = totalStakedFor(_accountAddress).sub(accounts[_accountAddress].totalLocked);
+
+        return unlockedTokens;
+    }
+
+    function _getTotalLockedOf(address _accountAddress) internal view returns (uint256) {
+        return accounts[_accountAddress].totalLocked;
+    }
+
+    /**
+     * @notice Check if `_accountAddress`'s by `_lockManager` can be unlocked
+     * @param _accountAddress Owner of lock
+     * @param _lockManager Manager of the lock for the given account
+     * @return Whether given lock of given account can be unlocked
+     */
+    function _canUnlock(address _accountAddress, address _lockManager, uint256 _amount) internal view returns (bool) {
+        Lock storage lock_ = accounts[_accountAddress].locks[_lockManager];
+        require(lock_.allowance > 0, ERROR_LOCK_DOES_NOT_EXIST);
```

Comment:
I wouldn't check the allowance here, I guess I'm inclined to support changing the allowance besides what you have been locked

Reply:
As the error messages says, the allowance is used as a flag of existence for locks.

Reply:
I see, then I think it's safe to not allow decreasing the allowance if you still have something blocked

Reply:
Do you mean to not allow decreasing it below the locked amount, or to not allow decreasing at all?